### PR TITLE
kernel: mark time conversion functions as constexpr in C++14.

### PR DESCRIPTION
**Pull request by jmgao**
Allow C++ code to evaluate time base conversion routines at compile time
by marking them as constexpr where possible.

### CODE REVIEW

**Comment by carlescufi** [comment]
@jmgao is this fixing a bug? If you want it included in 2.2 you should file a bug and add a `Fixes #<bug number>` in the PR description. Otherwise this can be merged when the 2.3 merge window opens (and after it's approved of course).

**Comment by jmgao** [comment]
Nope, it's just an improvement.

**Review by pabigot — CHANGES_REQUESTED** [review]
Other than it's a bit clunky to define then undefine a specifier, and a C++ version nit below, I have no objections to this.  I would have expected a compiler to optimize constant expression calculations so it shouldn't be necessary.

File: include/sys/time_units.h
Comment (jmgao) [reply include/sys/time_units.h]: Whoops, I forgot I had made that change and did a build assuming it was fixed on C++11. It appears to be a giant pain to make this work in C++11 because += is prohibited, and it's been 6 years since 2014 anyway, so I dropped the change to this function and fixed the commit message.

**Review by pabigot — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Review by andyross — APPROVED** [review]
Nice.  For extra credit, please provide a patch to gcc that allows constexpr as a feature in C11 code, so we can use these things in static initializers in Zephyr too. :)

(That loss was one of the bigger downsides to this feature, which otherwise I'm pretty proud of.)

**Review by tyrantlizard77 — APPROVED** [review]

**Pull request merged**
